### PR TITLE
Let resource client use timeout hints

### PR DESCRIPTION
**Pull request by brickZA**
This is an evil idea we hatched at CAM to enable devices to tell clients which requests are likely to be slow. Still experimenting, but if it looks nice we might try and propose it as a KATCP v5.1 protocol feature :)

### CODE REVIEW

File: katcp/test/test_inspecting_client.py [hunk 1]
```diff
@@ -155,6 +157,88 @@ def test_update_index(self):
...
 
+class TestInspectingClientInspect(tornado.testing.AsyncTestCase):
+    """Check that inspection populates the request/sensor index correctly"""
+
+    def setUp(self):
+        super(TestInspectingClientInspect, self).setUp()
+        self.server = DeviceTestServer('', 0)
+        start_thread_with_cleanup(self, self.server, start_timeout=1)
+        self.server_with_timeout_hints = DeviceTestServerWithTimeoutHints('', 0)
+        start_thread_with_cleanup(self, self.server_with_timeout_hints,
```
Comment (ajoubertza) [inline katcp/test/test_inspecting_client.py]: You don't seem to be using the two servers instantiated here.  The tests use instances created by `_get_server`.

File: katcp/test/test_inspecting_client.py [hunk 2]
```diff
@@ -155,6 +157,88 @@ def test_update_index(self):
...
+        hints = getattr(server, 'request_timeout_hints', {})
+        expected = {}
+        for req, handler in server._request_handlers.items():
+            expected[req] = {'name': req,
+                             'description': handler.__doc__,
+                             'timeout_hint': hints.get(req)}
+        return expected
+
+    @tornado.gen.coroutine
+    def _test_inspect_requests_no_timeout_hints(self, timeout_hints):
```
Comment (ajoubertza) [inline katcp/test/test_inspecting_client.py]: This name isn't so clear to me, since it might use timeout hints, depending on the parameter.  Suggest shortening to `_test_inspect_requests`.

File: katcp/test/test_resource_client.py [hunk 3]
```diff
@@ -73,18 +73,22 @@ class test_KATCPClientResourceRequest(unittest.TestCase):
...
+        self.assertEqual(self.DUT.timeout_hint, 33.34)
         # Check that we are registered to the correct ABC
         self.assertIsInstance(self.DUT, resource.KATCPRequest)
 
     def test_request(self):
         reply = self.DUT('parm1', 2)
         self.mock_client.wrapped_request.assert_called_once_with(
-            'the-request', 'parm1', 2)
+            'the-request', 'parm1', 2, timeout=33.34)
         self.assertIs(reply, self.mock_client.wrapped_request.return_value)
```
Comment (ajoubertza) [inline katcp/test/test_resource_client.py]: Would it be useful to add another test or test class like `test_KATCPClientResourceRequest`, that uses a timeout hint of `None`, just to check the timeout parameter comes through as expected?

File: katcp/inspecting_client.py [hunk 4]
```diff
@@ -621,12 +623,23 @@ def inspect_requests(self, name=None, timeout=None):
...
+            timeout_hints = {}
 
         requests_old = set(self._requests_index.keys())
         requests_updated = set()
         for msg in informs:
             req_name = msg.arguments[0]
-            req = {'description': msg.arguments[1]}
+            req = {'name': req_name,
+                   'description': msg.arguments[1],
+                   'timeout_hint': timeout_hints.get(req_name)}
```
Comment (brickZA) [inline katcp/inspecting_client.py]: Yes, that is about the shape of it. I'm requiring the dict key to always be present in the KATCPRequest stuff just to simplify the logic.

File: katcp/inspecting_client.py [hunk 5]
```diff
@@ -636,6 +649,53 @@ def inspect_requests(self, name=None, timeout=None):
             raise Return(AttrDict(added=added, removed=removed))
 
     @tornado.gen.coroutine
+    def _get_request_timeout_hints(self, name=None, timeout=None):
+        """Get request timeout hints from device
+
+        Parameters
+        =========
+        name : str or None, optional
+            Name of the request or None to get all request  timeout hints.
```
Comment (brickZA) [inline katcp/inspecting_client.py]: Done, commit pending.

File: katcp/resource.py [hunk 6]
```diff
@@ -401,8 +414,9 @@ def __call__(self, *args, **kwargs):
         Keyword Arguments
         -----------------
         timeout : None or float, optional
-            Timeout in seconds for the request. If None, use default for the
-            :class:`KATCPResource` instance that contains the request.
+            Timeout in seconds for the request. If None, use request timeout
+            hint recieved from server or default for the :class:`KATCPResource`
+            instance that contains the request if no hint is available.
```
Comment (brickZA) [inline katcp/resource.py]: Done, commit pending.

File: katcp/test/test_inspecting_client.py (hunk 1 above)
Comment (brickZA) [inline katcp/test/test_inspecting_client.py]: indeed, that seems to be the case. Fixed in pending commit.

File: katcp/test/test_inspecting_client.py (hunk 2 above)
Comment (brickZA) [inline katcp/test/test_inspecting_client.py]: yes, think what was a refactoring-non-renamo ;)

Fixed, commit pending.

**Comment by ludwigschwardt** [comment]
What is a timeout hint? Maybe update the PR title and description or even the ticket for us ignoramuses :-)

**Comment by brickZA** [comment]
@ajoubertza Changes pushed and ready to judgement

@ludwigschwardt The issue number is actually in the branch name (CB-569). It is an evil idea we hatched at CAM to enable devices to tell clients which requests are likely to be slow. Still experimenting, but if it looks nice we might try and propose it as a KATCP v5.1 protocol feature :)

File: katcp/test/test_resource_client.py (hunk 3 above)
Comment (brickZA) [inline katcp/test/test_resource_client.py]: added.

File: katcp/test/test_inspecting_client.py [hunk 7]
```diff
@@ -228,11 +220,11 @@ def _test_inspect_requests_no_timeout_hints(self, timeout_hints):
...
     @tornado.testing.gen_test
     def test_inspect_request_no_timeout_hints(self):
-        yield self._test_inspect_requests_no_timeout_hints(timeout_hints=False)
+        yield self._test_inspect_requests(timeout_hints=False)
 
     @tornado.testing.gen_test
     def test_inspect_request_with_timeout_hints(self):
-        yield self._test_inspect_requests_no_timeout_hints(
+        yield self._test_inspect_requests(
             timeout_hints=True)
```
Comment (ajoubertza) [inline katcp/test/test_inspecting_client.py]: Subjective, but this would be nicer to read on one line.  It isn't that long.

Comment (brickZA) [inline katcp/test/test_inspecting_client.py]: Doneded.

**Comment by ludwigschwardt** [comment]
Have you read issue [CB-569](https://skaafrica.atlassian.net/browse/CB-569)? :-)

**Comment by brickZA** [comment]
@ludwigschwardt Grandted, the issue is a bit dilatory. But if you read the documentation for the ?request-timeout-hint handler in server.py I think you'll get the picture of what we're trying to do :)

**Comment by ludwigschwardt** [comment]
So if the hint is accurate we basically have a blocking system? :-)
